File: 2021-11/2021-11-5 1218.最长定差子序列/LongestSubsequence.cs
public class Solution {
    public int LongestSubsequence(int[] arr, int difference) {
        var result = 0;
        // dp[i]
        var dp = new Dictionary<int, int>();
        foreach (var i in arr)
        {
            if (dp.TryGetValue(i - difference, out var pre))
            {
                dp[i] = pre + 1;
            }
            else
            {
                dp[i] = 1;
            }
            result = Math.Max(result,dp[i]);
        }

        return result;
    }
}

File: 2021-11/2021-11-5 1218.最长定差子序列/longestSubsequence.go
package main

func main() {

}

func longestSubsequence(arr []int, difference int) int {
	var temp int
	hash := make(map[int]int, 0)
	for _, v := range arr {
		hash[v] = hash[v-difference] + 1
		temp = max(temp, hash[v])
	}
	return temp
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
